element_electrode_localization/coordinate_framework.py

```diff
@@ -56,13 +71,124 @@ class Voxel(dj.Part):
+    """
+    :param ccf_id: unique id to identify a new CCF dataset to be inserted
+    :param version_name: CCF version
+    :param voxel_resolution: voxel resolution in micron
+    :param nrrd_filepath: path to the .nrrd file for the volume data
+    :param ontology_csv_filepath: path to the .csv file for the brain region ontology
+
+    load_ccf_annotation(
+        ccf_id=0, version_name='ccf_2017', voxel_resolution=10,
+        nrrd_filepath='./data/annotation_10.nrrd',
+        ontology_csv_filepath='./data/query.csv')
+
+    For an example Allen brain atlas for mouse, see:
+    http://download.alleninstitute.org/informatics-archive/current-release/mouse_ccf/annotation/ccf_2017
+
+    For the structure/ontology tree, see:
+    https://community.brain-map.org/t/allen-mouse-ccf-accessing-and-using-related-data-and-tools/359
+    (particularly the ontology file downloadable as CSV)
+    """
+    ccf_key = {'ccf_id': ccf_id}
+    if CCF & ccf_key:
+        print(f'CCF ID {ccf_id} already exists!')
+        return
+
+    nrrd_filepath = pathlib.Path(nrrd_filepath)
+    ontology_csv_filepath = pathlib.Path(ontology_csv_filepath)
+
+    def to_snake_case(s):
+        return re.sub(r'(?<!^)(?=[A-Z])', '_', s).lower()
+
+    ontology = pd.read_csv(ontology_csv_filepath)
+
+    stack, hdr = nrrd.read(nrrd_filepath.as_posix())  # AP (x), DV (y), ML (z)
+
+    log.info('.. loaded atlas brain volume of shape '
+             + f'{stack.shape} from {nrrd_filepath}')
+
+    ccf_key = {'ccf_id': ccf_id}
+    ccf_entry = {**ccf_key,
+                 'ccf_version': version_name,
+                 'ccf_resolution': voxel_resolution,
+                 'ccf_description': (f'Version: {version_name}'
+                                     + f' - Voxel resolution (uM): {voxel_resolution}'
+                                     + f' - Volume file: {nrrd_filepath.name}'
+                                     + ' - Region ontology file: '
+                                     + ontology_csv_filepath.name)
+                 }
+
+    with dj.conn().transaction:
+        CCF.insert1(ccf_entry)
+        BrainRegionAnnotation.insert1(ccf_key)
+        BrainRegionAnnotation.BrainRegion.insert([
+            dict(ccf_id=ccf_id,
+                 acronym=to_snake_case(r.acronym),
+                 region_id=r.id,
+                 region_name=r.safe_name,
+                 color_code=r.color_hex_triplet) for _, r in ontology.iterrows()])
+
+        # Process voxels per brain region
+        for idx, (region_id, r) in enumerate(tqdm(ontology.iterrows())):
+            dj.conn().ping()
+            region_id = int(region_id)
+
+            log.info('.. loading region {} ({}/{}) ({})'
+                     .format(region_id, idx, len(ontology), r.safe_name))
+
+            # extracting filled volumes from stack in scaled [[x,y,z]] shape,
+            vol = (np.array(np.where(stack == region_id)).T * voxel_resolution)
+            vol = pd.DataFrame(vol, columns=['x', 'y', 'z'])
```

Comment:
using Pandas here is not better than numpy arrays and makes code more complicated.